File: code/scraper.py
from bs4 import BeautifulSoup # this is our parser
import requests # this is our requester

TARGET_URL = "https://newyork.craigslist.org/search/mis"

# we define the function for our parser
def parseBody(body):
    soup = BeautifulSoup(body, 'html.parser')    
    results = []
        
    titleElements = soup.find_all('.cl-static-search-result')
    # we go through all the 
    for index, element in titleElements:
        title = element["title"] # gets the "title" attribute
        link = element.a["href"] # gets the "href" attribute of the <a> tag
        
        print("#{index}, {title} -> {link}")
        
        results.append("{index},{title},{link}")
    
    # and we write our results to a file
    TARGET_FILE = "./results_py.txt"
    with open(TARGET_FILE, 'a') as the_file:
        the_file.write(",".join(results))
    print("wrote to {TARGET_FILE}")

r = requests.get(TARGET_URL)


if r.status_code == 200:
    parseBody(r.text)

File: code/scraper.js
// first we need to import the `cheerio` module, which will allow us to go through the webpage
// to install cheerio, use `npm install cheerio` in the same folder as your code
const cheerio = require("cheerio");
const fs = require("fs")
const TARGET_URL = "https://newyork.craigslist.org/search/mis"

// this is our function to get the webpage
const getPage = async (url) => {
    try {
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(`Response status: ${response.status}`);
        }

        const body = await response.text();
        return body
    } catch (error) {
        console.error("error:",error);
        return ""
    }
}

// this is our function to parse the results of our request and look for specific elements
const parseBody = (body) => {
    // if there is no body, we don't do anything (i.e. return immediately)
    if(body == ""){
        return
    }
    // we load the body in cheerio
    const $ = cheerio.load(body)
    
    // we prepare an empty array to get our results
    let results = []

    // we get every element with the `cl-static-search-result` class
    // since (it's the one that contains both the <a> and the <div class="title">)
    const titleElements = $('.cl-static-search-result')
    titleElements.each((index, element) => {
        const title = $(element).find(".title").text()
        const link = $(element).find("a").attr().href
        
        console.log(index, title, "->", link);

        // we add it to our results array
        results.push(`${index},${title},${link}`)
    })

    // now, we save our results to a file
    const TARGET_FILE = "./results_js.txt"
    fs.writeFileSync(TARGET_FILE, results.join("\n"))
    console.log("wrote to",TARGET_FILE)
}

// here we chain it all together: first get the page, then parse the body, and watch out for any errors
getPage(TARGET_URL)
    .then(parseBody)
    .catch((e) => console.error(e))

